### ISSUE TITLE
VolksForth issue with CONST/CONIN

### ISSUE THREAD

**Issue by cstrotm** [nihirash/Agon-CPM2.2#11]
We (@uho and @cstrotm) are trying to get VolksForth CP/M (https://github.com/forth-ev/VolksForth/tree/master/8080/CPM) to run on Agon-CPM2.2.

VolksForth works on a wide range of real CP/M hardware and emulators.

On Agon-CPM2.2, it starts, prints out the startup banner, and then gets into problems when reading a key (or checking the status of the keyboard). It seems that Agon-CPM2.2 delivers an endless stream of keystrokes, which crashes the program.

The code executed can be found at 

https://github.com/forth-ev/VolksForth/blob/bda78c41f9b559bb5877300c2f065bb673d9b553/sources/cpm/source.fth#L2028

and

https://github.com/forth-ev/VolksForth/blob/bda78c41f9b559bb5877300c2f065bb673d9b553/sources/cpm/source.fth#L2052

@nihirash Are you aware of any incompatibilities of the BIOS CONST/CONIN interface that we might be able to work around in our code?

**Comment by nihirash** [comment]
CONST and CONIN working as expected for CP/M 2.2. 

I've tested it with ZMP, Kermit, Wordstar, Inforcom games and many other software without issues. All software works fine(ZMP and Kermit also switches IOByte and working with UART via same routines). 

I'll check behavior with this Forth implementation - if there way to make system more compatible - I'll fix it.

**Comment by cstrotm** [comment]
Thanks for checking. I will debug and maybe change the implementation in Forth to make it more robust. I will let you know about my finding, please keep this issue open for some time. I'll report back.

**Comment by uho** [comment]
Hallo Carsten,

es gibt ja die Version für CP/M+, die nicht mehr das BIOS verwendet. Die sollten wir auch probieren. Mein Agon-Board ist angekommen aber ich kann mich erst in der Woche vom 24. Juli darum kümmern.

Viele Grüße,
	Ulli

**Comment by uho** [comment]
There is a version of volksForth that runs on CP/M+ and does not use BIOS for accessing the keyboard. We'll try that.

Regards,
	Ulrich

**Comment by nihirash** [comment]
Just checked VolksForth - it works from my dev. source tree without any issues :-)

I'll prepare release update soon and publish it!

**Comment by nihirash** [comment]
Release version works fine with VolksForth. 

I even included distro as drive F. 
Please, @uho and @cstrotm - check latest version.

**Comment by uho** [comment]
Hi @nihirash, hi @cstrotm,

I just checked that volksForth on F: works with the new version. 

Somewhat. 

Using the new Version with MOS 1.03 was unstable, crashing sporadically. Even CP/M PIP did not work.

Updating to MOS 1.04 RC 1 using the [flash update procedure linked to in the Agon-CPM2.2 release notes](https://github.com/nihirash/Agon-CPM2.2/releases/tag/1.1) actually **bricked my Agon light 2**. :-)
Would have been a prior VDP firmware update be necessary?

I now have to investigate in bare metal recovery as described in [https://github.com/envenomator/agon-vdpflash](https://github.com/envenomator/agon-vdpflash)... Hopefully I find time to fix it.

**Comment by nihirash** [comment]
I'm using fresh VDP(from GIT) but looks like it should work with latest official VDP version. 

It's very strange that this MOS bricked your Agon, cause it contains too small count of changes from 1.03 version.

**Comment by uho** [comment]
I was able to restore MOS 1.03 and VDP 1.03 so I'm on save ground again. Good to know I can revert to this version at any time.

I retried using MOS 1.04 RC1 along with VDP 1.04 RC1 from the [agon-vdp repository](https://github.com/breakintoprogram/agon-vdp) but get the same effect as before:  

> After `flash mos.bin 36aa1177` flashing seems to be fine. The following automatic reboot shows just the MOS version line. A hardware reset (reset button / power cycle) then does no longer boot the system - just blank screen.

Any idea how to diagnose the issue? This is an Agon Light 2 Rev.B board from Olimex.

I stick to MOS/VDP 1.03 for now. volksForth works fine with the [previous Agon-CPM2.2-version](https://github.com/nihirash/Agon-CPM2.2/tree/914490da9ad29c17ef4ca98d5f01fb3a410e4357). I brought in the files as drive D:

It's unclear what initially caused our original problem. I cannot reproduce it with my setup here.

From my side this issue can be closed for now.
